### ISSUE TITLE
v4 passes through unknown HTML attribute to the DOM

### ISSUE THREAD

**Issue by shiningjason** [styled-components/styled-components#2131]
## Environment
npm packages:
1. react: v16.5.2
2. react-dom: v16.5.2
3. styled-components: v4.0.2

## Reproduction
https://codesandbox.io/s/zn5vrk0vpm

## Expected Behavior
https://codesandbox.io/s/20xlp242pr
There is no unknown-prop warning if I'm using `styled(styled.button).attrs({ customProp })` with styled-components v3.4.10.

## Actual Behavior
There is unknown-prop warning with styled-components v4.
<img width="961" alt="2018-10-20 11 33 29" src="https://user-images.githubusercontent.com/652135/47250898-f22e4b00-d45c-11e8-8e64-dd29534dea1e.png">

If this is not a bug, and there is something missing. Please let me know, thanks!

**Comment by Fer0x** [comment]
Don't think that it is expected behavior.

Also example from docs https://www.styled-components.com/docs/api#attrs works in similar way:

![image](https://user-images.githubusercontent.com/110616/47260819-0dd03a80-d4cb-11e8-8b3d-33120ac64402.png)

**Comment by mxstbr** [comment]
This is definitely a bug, any ideas why this broke @probablyup?

**Comment by Fer0x** [comment]
@shiningjason @mxstbr 

What's interesting is that styled-components@3.x works much the same but instead of camelCase props are inserted in lowercase and React doesn't warn about this.

![image](https://user-images.githubusercontent.com/110616/47265731-29702b00-d535-11e8-8778-aba45a06d6f0.png)

**Comment by mxstbr** [comment]
Oh, that's even more strange 🤔

**Comment by Fer0x** [comment]
Fixed in #2133
However, pay attention to #2093. It will add new way to tell which attributes should pass as html attributes

**Comment by mbrowne** [comment]
I'm seeing a similar issue with the new `as` prop, e.g.:

```js
...
import { Link } from 'react-router-dom';

const StyledLink = styled.a`
    ${p => (p.linkStyle === 'upperCase' ? 'text-transform: uppercase' : '')}
`

<StyledLink as={Link} linkStyle="upperCase" to="/">test link</StyledLink>
```

> Warning: React does not recognize the `linkStyle` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `linkstyle` instead. If you accidentally passed it from a parent component, remove it from the DOM element.

Is this related to an existing issue or should I open a new issue?

**Comment by probablyup** [comment]
@mbrowne Since `Link` is a custom component, we don't do prop filtering and pass all props through to it. Until we settle on one of the several prop handling proposals floating around for v5 I don't think we're going to have a good solution for this.

**Comment by byzyk** [comment]
Is there any ETA on when this can be addressed? Or let me know if I can help with the PR and what the challenge is.

**Comment by byzyk** [comment]
As a workaround for eliminating React warnings in custom components, I had to make use of `data-*` attributes.

**Comment by jarodtaylor** [comment]
Will this be resolved in v5 for the `Link` (or any other custom components)?

**Comment by dustwise** [comment]
Any update on this now that v5 beta is out? I'm attempting to use Formik's Field component with styled-components and want to customize the css if an error is present or not.

**Comment by srikanthbhamidipalli** [comment]
Still I'm getting error.. :cry:

**Comment by GuilleDF** [comment]
Seeing as this still happens in v5 for custom components, can we reopen this issue? Or create a new one?

**Comment by probablyup** [comment]
This isn’t a bug, we explicitly do not filter props for custom components because we don’t know if the component needs to consume the prop as well.

**Comment by ItsMrAkhil** [comment]
@probablyup @mxstbr Why can't we filter properties based on some prefix like `$`, like below?

`<StyledComp $someStyleProp={someValue} />`

The `$someStyleProp` will not be passed down as the user explicitly wants to ignore the property being passed down?

**Comment by probablyup** [comment]
I proposed this a while ago actually. We could, but if another library
decides to do the same it’d be a mess basically.

**Comment by Maushundb** [comment]
The top answer [here](https://stackoverflow.com/questions/49834251/how-to-extend-styled-component-without-passing-props-to-underlying-dom-element) solved it for me
